clio-client/src/main/scala/org/broadinstitute/clio/client/dispatch/AddExecutor.scala

```diff
@@ -42,8 +46,106 @@ class AddExecutor[TI <: TransferIndex](addCommand: AddCommand[TI])
       Future
         .failed(_) logErrorMsg s"Metadata at $location cannot be added to Clio", {
         decoded =>
-          webClient.upsert(addCommand.index)(addCommand.key, decoded)
+          {
+            for {
+              existingMetadata <- queryForKey(webClient).logErrorMsg(
+                s"Could not query the $name. No files have been deleted."
+              )
+              upsertResponse <- addFiles(
+                webClient,
+                decoded,
+                existingMetadata,
+                addCommand.forceUpdate
+              ).fold(
+                left => Future.failed(left) logErrorMsg left.getMessage,
+                right => right
+              )
+            } yield {
+              upsertResponse
+            }
+          }
       }
     )
   }
+
+  private def addFiles(
+    client: ClioWebClient,
+    newMetadata: addCommand.index.MetadataType,
+    existingMetadata: Option[addCommand.index.MetadataType],
+    forceUpdate: Boolean
+  )(implicit ec: ExecutionContext): Either[Throwable, Future[UpsertId]] = {
+    val map = existingMetadata.map(diffMetadata(_, newMetadata)).getOrElse(Map.empty)
+    for {
+      upsertResponse <- Either.cond(
+        map.isEmpty || forceUpdate,
+        client
+          .upsert(addCommand.index)(addCommand.key, newMetadata)
+          .logErrorMsg(
+            s"""An error occurred while adding the $name record in Clio.""".stripMargin
+          ),
+        new RuntimeException(
+          s"Adding this document that will overwrite the following existing metadata: " +
+            map.keys.mkString(", ") +
+            s" Use the '--force-update true' to overwrite this data."
+        )
+      )
+
+    } yield {
+      upsertResponse
+    }
+  }
+
+  private def diffMetadata(
+    existingMetadata: addCommand.index.MetadataType,
+    newMetadata: addCommand.index.MetadataType
+  ): Map[String, Any] = {
+    val storedFields =
+      new CaseClassMapper[addCommand.index.MetadataType].vals(existingMetadata)
+    val upsertFields =
+      new CaseClassMapper[addCommand.index.MetadataType].vals(newMetadata)
+
+    (upsertFields.toSet diff storedFields.toSet).toMap
+  }
+
+  private def queryForKey(
```

Comment:
We've got variations of this logic in the add, move, and delete CLPs... could we take the chance to move it into the web-client? Or better to split off as a tech-debt ticket?

Reply:
Yeah I started doing this but it ended up being a little more work and seemed outside the scope of the ticket. I'll take another stab at it.